compiler/rustc_ast_lowering/src/expr.rs

```diff
@@ -76,7 +76,12 @@ impl<'hir> LoweringContext<'_, 'hir> {
                 ExprKind::Array(exprs) => hir::ExprKind::Array(self.lower_exprs(exprs)),
                 ExprKind::ConstBlock(c) => {
                     self.has_inline_consts = true;
-                    hir::ExprKind::ConstBlock(self.lower_expr(c))
+                    self.with_new_scopes(e.span, |this| {
+                        let coroutine_kind = this.coroutine_kind.take();
```

Comment:
Kind of strange that `coroutine_kind` is not handled in `with_new_scopes`. Is there some other `with_body_*`-like function that does the coroutine kind handling too?

Reply:
I have been looking into this in a branch, but didn't want to add more stuff to this PR.

we can get it all working by merging `with_new_scopes` into `lower_body`, but it needs some managing of the right spans to maintain diagnostic quality.